[KhoonSun47] iP

KhoonSun47 opened the pull request:
## William

>"First, solve the problem. Then, write the code." - John Johnson --> [Source](https://gist.github.com/scottashipp/c587dd35d253ce7c6a2bbee15abe143c)

William frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/KhoonSun47/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you, it is that simple!!! 😉

And it is **FREE!**

Features:

 - [x] Managing tasks such as Add Task, View Task, etc
 - [ ] Managing deadlines (coming soon!)
 - [ ] Reminders (coming soon!)
 ---
 If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
 ``` java
 public static void main(String[] args) {
     new William(filePath).run();
 }
 ```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

KhoonSun47 committed:
Level 0. Rename, Greet, Exit.

KhoonSun47 committed:
Level 1. Echo

KhoonSun47 committed:
Level 2. Add, List

KhoonSun47 committed:
Level 3. Mark as Done.

KhoonSun47 committed:
Level 4: ToDos, Events, Deadlines

Add Deadline, Event and Todo classes
William.java: Move existing and new methods to Methods.java
Commands.java: Add new enums

KhoonSun47 committed:
A-TextUiTesting: Automated Text UI Testing

William.java and Methods.java: Remove tab space
runtest.sh: Update javac and java commands to match the name of main class
input.txt: Add input test cases
EXPECTED.txt: Add expected output test cases

KhoonSun47 committed:
Level-5: Handle Errors

Add WilliamException.java
Methods.java and William.java: Add try-catch to handle incorrect inputs for the three type of tasks

KhoonSun47 committed:
Level-6: Delete

Commands.java: Add new command delete
Methods.java: Add new method for deleting tasks from the list
William.java: Add new switch case for delete

KhoonSun47 committed:
A-Enums: Use Enums

KhoonSun47 committed:
Level 7: Save

Add FileSaving.java and data folder which contains tasks.txt
Commands.java, Methods.java: Update JavaDocs comment
Commands.java, Methods.java, Task.java: Update imported classes to be listed explicitly
Todo.java, Event.java, Deadline.java, Task.java: Add new constructor and new methods to be used by FileSaving class

KhoonSun47 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Level 7: Save

KhoonSun47 committed:
Level 8: Dates and Times

FileSaving.java, William.java: Modify variables that is related to date into LocalDateTime
Task.java, Event.java, Deadline.java: Modify variables that is related to date into LocalDateTime and create new methods to handle LocalDateTime
Methods.java: Add and modify methods and variables to handle LocalDateTime

KhoonSun47 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Level 8: Dates and Times

KhoonSun47 committed:
A-MoreOOP: Use more OOP

Create commands, exceptions, tasks and utilities folder
Move Commands.java into commands
Move WilliamException java into exceptions
Move Deadline.java, Event.java, Task.java, Todo.java into tasks
Created AdditionalInfoParser, DateAndTimeParser, Parser, TaskList and Ui files
Rename SavingFile.java as Storage.java
Delete Methods.java and move the methods into other Java files based on functionality
Update William.java to make use of the new classes
Add JavaDocs to the new classes

KhoonSun47 committed:
Merge branch 'A-MoreOOP'

* A-MoreOOP:
  A-MoreOOP: Use more OOP

KhoonSun47 committed:
A-Packages: Organize into Packages

William.java: Add comment to indicate that the classes have already been divided into packages during the A-MoreOOP increment

KhoonSun47 committed:
Revert "A-Packages: Organize into Packages"

This reverts commit 09fd267a0f84b4d27904e323e65157c4ad81046f.

KhoonSun47 committed:
A-Packages: Organize into Packages

William.java: Indicate that the classes have been organised into suitable java packages during A-MoreOOP increment

KhoonSun47 committed:
Merge branch 'A-Packages'

* A-Packages:
  A-Packages: Organize into Packages

KhoonSun47 committed:
Merge commit 'a6f73244a20539e20c2189973ebf597df34ef631'

* commit 'a6f73244a20539e20c2189973ebf597df34ef631':
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

KhoonSun47 committed:
A-Gradle: Use A-Gradle

build.gradle: Update the main class to "William" and enable assertions

KhoonSun47 committed:
A-JUnit: Add JUnit Tests

Add StorageStub.java and TaskListStub.java to mimic Storage and TaskList class to test Parser.java in isolation
Add DateAndTimeParserTest.java which contains test method for testing methods of the DateAndTimeParser class
Add ParserTest.java which contains test method for testing methods of the Parser class
AdditionalInfoParser.java: Change the checkAdditionalDetailEmpty method for strings that contains only whitespace character

KhoonSun47 committed:
Merge branch 'A-JUnit'

* A-JUnit:
  A-JUnit: Add JUnit Tests

KhoonSun47 committed:
A-JavaDoc: JavaDoc

William.java, Commands.java, WilliamException.java, AdditionalInfoParser.java, Parser.java, Storage.java, StorageStub.java, TaskList.java, TaskListStub.java, Ui.java, DateAndTimeParserTest.java, ParserTest.java: Update JavaDocs for class and methods based on the Java Standards

KhoonSun47 committed:
A-Coding Standard: Follow the Coding Standard

William.java, WilliamException.java, Deadline.java, Event.java, Task.java, Todo.java, AdditionalInfoParser.java, DateAndTimeParser.java, Parser.java, Storage.java, StorageStub.java, TaskList.java, TaskListStub.java, Ui.java, DateAndTimeParserTest.java, ParserTest.java: Change constant names to be all uppercase using underscore to separate words, change indentation to be 4 spaces, change the switch case to have no indentation for case clauses

KhoonSun47 committed:
Level-9: Find

Commands.java: Add new command 'find'
Parser.java: Add new switch case clause for 'find'
TaskList.java: Add new method to find tasks that matches the input
Parser.java, StorageStub.java, TaskList.java, TaskListStub.java, ParserTest.java: Change the ordering of import statement

KhoonSun47 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  A-JavaDoc: JavaDoc

KhoonSun47 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  A-Coding Standard: Follow the Coding Standard

# Conflicts:
#	src/main/java/William.java
#	src/main/java/exceptions/WilliamException.java
#	src/main/java/utilities/AdditionalInfoParser.java
#	src/main/java/utilities/Parser.java
#	src/main/java/utilities/Storage.java
#	src/main/java/utilities/StorageStub.java
#	src/main/java/utilities/TaskList.java
#	src/main/java/utilities/TaskListStub.java
#	src/test/java/utilities/DateAndTimeParserTest.java
#	src/test/java/utilities/ParserTest.java

KhoonSun47 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Level-9: Find

# Conflicts:
#	src/main/java/William.java
#	src/main/java/commands/Commands.java
#	src/main/java/utilities/Parser.java
#	src/main/java/utilities/Storage.java
#	src/main/java/utilities/StorageStub.java
#	src/main/java/utilities/TaskList.java
#	src/main/java/utilities/TaskListStub.java
#	src/test/java/utilities/ParserTest.java

KhoonSun47 committed:
A-CheckStyle: Use CheckStyle

Add checkstyle.xml, suppressions.xml as pre-requisite to check if the main/test code compiles with the style rules

build.gradle: Add relevant lines so that checkstyle can be used with gradle
Event.java: Change access modifier of toDate and fromDate to private
Task.java: Change the order of the static import
AdditionalInfoParser.java: Change the length of the lines to be 120 character per line
Parser.java: Change the indentation
Storage.java: Add a default clause in the switch case
StorageStub.java, TaskList.java, TaskListStub.java: Add spacing in between the imports

KhoonSun47 committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  A-CheckStyle: Use CheckStyle

headcube1 commented on src/main/java/utilities/Storage.java:
```diff
@@ -0,0 +1,119 @@
+        // Each part can contain any characters, with whitespace handling Capture non-whitespace
+        // characters in the first two parts and at least one character in the last three parts
+
+        String matchingPattern =
```
Line length of all your code is just nice, not too long!

headcube1 commented on src/main/java/William.java:
```diff
@@ -0,0 +1,50 @@
+ * packages: commands, exceptions, tasks and utilities)
+ */
+public class William {
+
```
Basic indentation is all 4 spaces, which follows the Java coding standard, Nice!

headcube1 commented on src/main/java/William.java:
```diff
@@ -0,0 +1,50 @@
+     * 
+     * @param filePath The file path
+     */
+    public William(String filePath) {
```
I like how you ensured all your booleans are named isXXX!

headcube1 commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.time.LocalDateTime;
+
+public class Deadline extends Task {
```
I like that you ensured the correct whitespace requirements!

headcube1 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,62 @@
+import java.time.format.DateTimeFormatter;
+
+public class Task {
+    private static final DateTimeFormatter INPUT_FORMAT =
```
Should there be spacing between your fields as logical units within a block should be separated by one blank line?

KhoonSun47 replied:
Great observation, I did not think of that. I will incorporate the required adjustments in the next commit!

headcube1 reviewed (commented):
Nice work!

KhoonSun47 committed:
A-Varargs: Use Varargs

AdditionalInfoParser.java: Use Varargs to create a general method for splitting the input, replacing two of the methods
Parser.java: Separate methods for switch clause
DateAndTimeParserTest.java: Change the ordering of the imports

KhoonSun47 committed:
Merge branch 'branch-A-Varargs'

* branch-A-Varargs:
  A-Varargs: Use Varargs

starsia commented on src/main/java/William.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public static void main(String[] args) {
+        new William(filePath).run();
```
Appreciate how neat your Main class is!

starsia commented on src/main/java/commands/Commands.java:
```diff
@@ -0,0 +1,8 @@
+ * Contains the list of commands to be used in the switch case for William.java
+ */
+public enum Commands {
+    list, bye, mark, unmark, todo, deadline, event, delete, find
```
Not sure if the values in the ENUM type should be capitalised, since they are constant!

<img width="819" alt="image" src="https://github.com/nus-cs2103-AY2324S2/ip/assets/104408800/7cb2eb13-5726-4eeb-b0ca-27d3a58ff7ad">

starsia commented on src/main/java/exceptions/WilliamException.java:
```diff
@@ -0,0 +1,10 @@
+package exceptions;
```
I like the foresight to have a package exclusively for exceptions. Definitely useful as the project gets bigger.

starsia commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public String[] getTimes() {
+        return new String[] {super.changeDateToString(byDate), "NILL"};
```
Did you mean NIL?

starsia commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + super.outputDateAsString(fromDate) + " to: "
```
I like that you had single spaces between operators, which makes it easy to read!

starsia commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,62 @@
+        this.isDone = isDone;
+    }
+
+    public void changeIsDone() {
```
Great use of the XOR function!

starsia commented on src/main/java/utilities/TaskList.java:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Constructor if there is an existing txt file with tasks
+     */
+    public TaskList(List<Task> tasks) {
```
Would it be more specific to constrain the List<Task> type to ArrayList<Task>? Instead of the abstract class. This may prevent other List data structures like Queue/Stack to be passed into the function, which could interfere with the intended outcome of any add/remove/search methods you have.

starsia commented on src/test/java/utilities/ParserTest.java:
```diff
@@ -0,0 +1,165 @@
+     * been marked successfully
+     */
+    @Test
+    public void parseCommands_markCommand_markTaskAsNotDone() {
```
I like your comments for the testing of the Parser and DateAndTime classes!

starsia reviewed (commented):
Coding standard is great overall! Just a note or two on Enums and nits. 👍

KhoonSun47 committed:
Level-10: GUI

Add a new folder javafx that contains DialogBox.java, Main.java, MainWindow.java
Add a new folder primary that contains William.java

Add a new folder resources, which contains 2 other folders: images and view
Add two images: luffy.jpeg and chopper.jpeg into images
Add DialogBox.fxml and MainWindow.fxml into view

Create Launcher.java to start the JavaFX application
Edit Parser.java, Ui.java, TaskList.java such that the chatbot output goes into the JavaFX application, rather than in terminal

Issues that needs to be clarified:
Need to resolve the issue when retrieving JPEG files, the given method does not work, now currently using absolute path
Need to resolve the issue when retrieving FXML files, the given method does not work, now currently using absolute path

KhoonSun47 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Level-10: GUI

KhoonSun47 committed:
A-Assertions: Use Assertions and Refine Input Validation

- Incorporate Java assert statements in Ui.java, AdditionalInfoParser.java,
  and DateAndTimeParser.java to document essential assumptions.

- Modify AdditionalInfoParser.java's splitInput method to enhance input
  validation, throwing an exception when required keywords are missing.
  This change ensures better input handling and clearer error feedback to users.

- These adjustments are in line with the strategy to use assertions for
  internal consistency checks rather than relying solely on JUnit assertions,
  aiming for improved code clarity and reliability.

KhoonSun47 committed:
A-Streams: Use Streams

TaskList class contains various methods that can be refactor to enhance code readability and maintainability using Java Streams API.

These are the following modifications:
- The printList method now uses streams to generate a list of tasks, making the code more concise.
- The findTasks method has been refactored to use streams for filtering tasks based on user input, then generating a list of matching tasks, streamlining the process of identifying these tasks.

Using of Java Streams API is preferable over the previous methods as it aim to improve the clarity of the TaskList class's operations, which could also help to facilitates modifications and extensions in the future.

KhoonSun47 committed:
Merge pull request #1 from KhoonSun47/branch-A-Streams

A-Streams: Use Streams

KhoonSun47 committed:
Merge branch 'master' of https://github.com/KhoonSun47/ip

* 'master' of https://github.com/KhoonSun47/ip:
  A-Streams: Use Streams

KhoonSun47 committed:
Merge branch 'master' into branch-A-Assertions

* master:
  A-Streams: Use Streams

KhoonSun47 committed:
Merge pull request #2 from KhoonSun47/branch-A-Assertions

Branch-A-Assertions

KhoonSun47 committed:
Merge branch 'master' of https://github.com/KhoonSun47/ip

* 'master' of https://github.com/KhoonSun47/ip:
  A-Assertions: Use Assertions and Refine Input Validation

KhoonSun47 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  A-Streams: Use Streams
  A-Assertions: Use Assertions and Refine Input Validation

KhoonSun47 committed:
Rename folder into ui, which contains the DialogBox.java, Main.java and MainWindow.java that is used for the chatbot UI

KhoonSun47 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Rename folder into ui, which contains the DialogBox.java, Main.java and MainWindow.java that is used for the chatbot UI

KhoonSun47 committed:
Remove ui folder which is a replicate of javafx folder

KhoonSun47 committed:
Added comment to aid to solve GUI Issues

KhoonSun47 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Added comment to aid to solve GUI Issues
  Remove ui folder which is a replicate of javafx folder

KhoonSun47 committed:
A-Code Quality: Improve Code Quality

Refactor code for improved quality and maintainability

- Refactor Commands.java to use uppercase for enum constants, aligning with Java naming conventions for constants.
- Make INPUT_FORMAT and OUTPUT_FORMAT in Task.java final, ensuring immutability of format strings.
- Update DialogBox.java, Main.java, and MainWindow.java to use relative paths for FXML and image resources, fixing resource loading issues for Level 10 increment.
- Reorganize project structure: create 'ui' folder, remove 'javafx' folder, and move relevant UI classes to enhance project organization.
- Refactor Storage.java to better adhere to the Single Level of Abstraction Principle (SLAP) by separating file loading logic into distinct methods.
- Update ParserTest.java to reflect changes in enum naming conventions to uppercase, ensuring consistency across the codebase.

KhoonSun47 committed:
Merge pull request #3 from KhoonSun47/branch-A-CodeQuality

Branch-A-CodeQuality

KhoonSun47 committed:
Merge branch 'master' of https://github.com/KhoonSun47/ip

* 'master' of https://github.com/KhoonSun47/ip:
  A-Code Quality: Improve Code Quality

KhoonSun47 committed:
BCD-Extension: C-Priority

Add Priorities.java that stores Priorities enum

Commands.java: Add new enum, Priority and Sort
Deadline.java, Event.java, Task.java, Todo.java: Add priority field
Parser.java: Add two new switch case clause, Priority and Sort, and a new method to handle priority
Storage.java: Update the storing and retrieving of details with new additional field priority
TaskList.java: Add two new methods to add priority to a task, and sort the list of tasks based on priority

KhoonSun47 committed:
Merge branch 'BCD-Extension'

* BCD-Extension:
  BCD-Extension: C-Priority

KhoonSun47 committed:
Update README.md

KhoonSun47 committed:
Update README.md with formatting syntax

KhoonSun47 committed:
Update README.md with new descriptions

KhoonSun47 committed:
Merge branch 'master' of https://github.com/KhoonSun47/ip

* 'master' of https://github.com/KhoonSun47/ip:
  Update README.md with new descriptions
  Update README.md with formatting syntax
  Update README.md

KhoonSun47 committed:
Update README.md for testing purposes

README.md: Updated the user guide such that it tallies with GitHub README

KhoonSun47 committed:
A-BetterGui: Better GUI

Update DialogBox.fxml: Change the border colour and padding for individual dialogs
Update MainWindow.fxml: Change the ScrollPane such that text does not get truncated, and UI for TextField and Button
Update Main.java and MainWindow.java: Change the product name to William
Update DialogBox.java: Change the format of profile picture

Update README.md: Add link to Ui.png
Add Ui.png into docs folder

KhoonSun47 committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  A-BetterGui: Better GUI

KhoonSun47 committed:
Update README.md with Ui.png

KhoonSun47 committed:
Merge branch 'master' of https://github.com/KhoonSun47/ip

* 'master' of https://github.com/KhoonSun47/ip:
  Update README.md with Ui.png

KhoonSun47 committed:
A-UserGuide: User Guide

Update README.md: Fix the formatting on table so that it display correctly on GitHub Page

KhoonSun47 committed:
A-MoreErrorHandling: More error handling

Update README.md: To inform about one of the error that could occur if change syntax of data file
Task.java, Parser.java, TaskList.java: Update the unmark/mark methods to handle errors
TaskListStub.java, DateAndTimeParserTest.java, ParserTest.java: Add new test cases
Storage.java: Update matching pattern to detect when there is additional field added to the data
Parser.java: Update handleBye method to quit the application

KhoonSun47 committed:
Merge branch 'branch-A-MoreErrorHandling'

* branch-A-MoreErrorHandling:
  A-MoreErrorHandling: More error handling

KhoonSun47 committed:
More Error Handling

Ui.java: Updated loadTasks method to create a directory if not found

KhoonSun47 committed:
A-AiAssisted: Enhance using AI tools

Ui.java: Add a comment to loadTask
Storage.java: Add a comment on how AI tool improve MATCHING_PATTERN
TaskList.java: Add a comment on how AI tool improve sort method using Stream API
